### PR TITLE
SW-688_PH-502_elastic_search_issue

### PR DESCRIPTION
**Pull request by dblaichinger**

### CODE REVIEW

File: src/Core/Content/Product/SalesChannel/Search/ProductSearchRoute.php [hunk 1]
```diff
@@ -101,6 +102,11 @@ public function load(
         SalesChannelContext $context,
         ?Criteria $criteria = null
     ): ProductSearchRouteResponse {
+
+        if (!Utils::versionLowerThan('6.4.0.0')) {
```
Comment (TobiasGraml11) [inline src/Core/Content/Product/SalesChannel/Search/ProductSearchRoute.php:106]: :thinking: What do you think about a utils function called something like `Utils::versionAboveOrEqual`, which represents (in this example) all versions starting from `6.4.0.0`.
It's not like this is some complex statement, but it would make it a bit more clear.

File: src/Core/Content/Product/SearchKeyword/ProductSearchBuilder.php [hunk 2]
```diff
@@ -71,8 +95,51 @@ public function build(Request $request, Criteria $criteria, SalesChannelContext
         );
     }
 
-    public function buildParent(Request $request, Criteria $criteria, SalesChannelContext $context): void
+
```
Comment (TobiasGraml11) [inline src/Core/Content/Product/SearchKeyword/ProductSearchBuilder.php:98]:
```suggestion
```

File: src/Core/Content/Product/SalesChannel/Search/ProductSearchRoute.php (hunk 1 above)
Comment (dblaichinger) [reply src/Core/Content/Product/SalesChannel/Search/ProductSearchRoute.php:106]: Agree, having a separate method for version greater or equal makes sense.

File: src/Core/Content/Product/SearchKeyword/ProductSearchBuilder.php [hunk 3]
```diff
@@ -71,8 +95,51 @@ public function build(Request $request, Criteria $criteria, SalesChannelContext
...
+                )
+            );
+        }
+        $criteria->addQuery(
+            new ScoreQuery(
+                new ContainsFilter('product.searchKeywords.keyword', $pattern->getOriginal()->getTerm()),
+                $pattern->getOriginal()->getScore(),
+                'product.searchKeywords.ranking'
+            )
+        );
```
[the remark this answers is not in the document]
Comment (dblaichinger) [reply src/Core/Content/Product/SearchKeyword/ProductSearchBuilder.php:128]: This was intended, as the methods are copied from SW:
* https://github.com/shopware/platform/blob/trunk/src%2FCore%2FContent%2FProduct%2FSearchKeyword%2FProductSearchBuilder.php

We need to overrite it, to get rid of the `MissingRequestParameterException` which prevents empty search. In case SW does some modifications again, we just need to add another method for the version and don't need to debug the combined methods.

**Review by TobiasGraml11 — APPROVED** [review]

File: tests/Core/Content/Product/SalesChannel/Listing/ProductSearchRouteTest.php [hunk 4]
```diff
@@ -175,4 +178,46 @@ public function testSearchingVariantProductNumberWillShowVariant(
         $searchedProduct = $result->first();
         $this->assertSame($expectedProductNumber, $searchedProduct->getProductNumber());
     }
+
+    public function testAddElasticSearchContextForShopware64AndGreater(): void
```
Comment (TheKeymaster) [inline tests/Core/Content/Product/SalesChannel/Listing/ProductSearchRouteTest.php:182]: Maybe rephrase this a bit
```suggestion
    public function testElasticSearchContextIsAddedForShopware64AndGreater(): void
```

File: tests/Core/Content/Product/SearchKeyword/ProductSearchBuilderTest.php [hunk 5]
```diff
@@ -6,38 +6,83 @@
...
+        parent::setUp();
+    }
+
+    public function testSuggestRouteIsIgnoredByFindologic(): void
+    {
         // Ensure Shopware handles the request.
-        $productSearchBuilderMock->expects($this->once())->method('buildParent');
+        $this->productSearchBuilderMock->expects($this->once())->method('buildParent');
+        $this->productSearchBuilderMock->expects($this->never())->method('buildShopware63AndLower');
+        $this->productSearchBuilderMock->expects($this->never())->method('buildShopware64AndGreater');
```
Comment (TheKeymaster) [inline tests/Core/Content/Product/SearchKeyword/ProductSearchBuilderTest.php:54]: Some of these expectations are pretty repetitive. Maybe you can move them into their own method such as `setProductSearchBuilderExpectations`.

File: tests/Core/Content/Product/SearchKeyword/ProductSearchBuilderTest.php [hunk 6]
```diff
@@ -6,38 +6,83 @@
...
 
         $request = Request::create('http://your-shop.de/suggest?search=blubbergurken');
         $criteria = new Criteria();
-        $salesChannelContext = $this->buildSalesChannelContext(Defaults::SALES_CHANNEL, 'http://test.de');
+        $this->productSearchBuilderMock->build($request, $criteria, $this->salesChannelContext);
+    }
+
+    public function testBuildMethodForShopwareLower64IsUsed(): void
+    {
+        if (Utils::versionGreaterOrEqual('6.4.0.0')) {
```
Comment (TheKeymaster) [inline tests/Core/Content/Product/SearchKeyword/ProductSearchBuilderTest.php:63]: In the other test you combined two test scenarios in one. I think it makes sense to merge them here aswell.

**Review by TheKeymaster — CHANGES_REQUESTED** [review]

File: src/Resources/views/storefront/layout/meta.html.twig [hunk 7]
```diff
@@ -2,7 +2,7 @@
 
 {% block layout_head_javascript_tracking %}
     {% if page.header.extensions.flConfig.active %}
-        {% if page.header.extensions.flConfig.integrationType === constant('FINDOLOGIC\\FinSearch\\Findologic\\IntegrationType::DI') %}
+        {% if page.header.extensions.flConfig.integrationType == constant('FINDOLOGIC\\FinSearch\\Findologic\\IntegrationType::DI') %}
```
[the remark this answers is not in the document]
Comment (dblaichinger) [reply src/Resources/views/storefront/layout/meta.html.twig:5]: No I found it in the log of staging.muenkel where I added the quick-fix to prevent flickering.

File: src/Storefront/Controller/SearchController.php [hunk 8]
```diff
@@ -143,8 +144,12 @@ public function filter(Request $request, SalesChannelContext $context): Response
     {
         $event = new ProductSearchCriteriaEvent($request, new Criteria(), $context);
         $this->findologicSearchService->doFilter($event);
-        $result = $this->filterHandler->handleAvailableFilters($event);
 
+        if (!Utils::isFindologicEnabled($context)) {
```
[the remark this answers is not in the document]
Comment (dblaichinger) [reply src/Storefront/Controller/SearchController.php:148]: True, since `CmsController::filter` is doing the same I'm going to move it there as well.

File: src/Resources/config/services/decorators.xml [hunk 9]
```diff
@@ -90,21 +90,21 @@
         </service>
 
         <service
```
[the remark this answers is not in the document]
Comment (dblaichinger) [reply src/Resources/config/services/decorators.xml:92]: What do you mean with missing? Shopware comes with [Shopware\\Elasticsearch\\](https://github.com/shopware/platform/blob/trunk/composer.json#L370) by default. It would need to get removed manually, right?

[the remark this answers is not in the document]
Comment (dblaichinger) [reply src/Resources/config/services/decorators.xml:92]: I've renamed the ElasticSearch directory and executed `composer remove shopware/elasticsearch` to match your scenario. Findologic search works, no error regarding undefined service etc.
Screen Shot 2022-03-15 at 10 45 52
.

**Review by TheKeymaster — APPROVED** [review]
